src/lib/xwalk-hack/churn.ts

```diff
@@ -0,0 +1,205 @@
+  // happen in response to a signalling channel message.
+  private haveTransformerConfig_ :(config:TransformerConfig) => void;
+  private onceHaveTransformerConfig_ = new Promise((F, R) => {
+    this.haveTransformerConfig_ = F;
+  });
+
+  constructor(rtcPc:freedom.RTCPeerConnection.RTCPeerConnection,
+              private name_ = 'unnamed-churn-' + Connection.id_,
+              private skipPublicEndpoint_?:boolean,
+              private portControl_?:freedom.PortControl.PortControl,
+              private preferredTransformerConfig_?:TransformerConfig) {
+    Connection.id_++;
+
+    this.signalForPeerQueue = new handler.Queue<ChurnSignallingMessage,void>();
+    this.configureConnection_(rtcPc);
+
+    // Forward onceXxx promises.
+    this.onceConnected = this.pc_.onceConnected;
+    this.onceClosed = this.pc_.onceClosed;
+  }
+
+  private configureConnection_ = (
+      freedomPc:freedom.RTCPeerConnection.RTCPeerConnection) => {
+    this.pc_ = new peerconnection.PeerConnectionClass(
+        freedomPc, this.name_);
+
+    // This will insert the caesar shift key
+    this.pc_.mungeLocalDescription = this.mungeLocalSdp_;
+
+    this.peerOpenedChannelQueue = this.pc_.peerOpenedChannelQueue;
+
+    this.pc_.signalForPeerQueue.setSyncHandler(
+        (message:signals.Message) => {
+      if (message.type === signals.Type.CANDIDATE) {
+        var c = Candidate.fromRTCIceCandidate(message.candidate);
+        if (c.protocol === 'udp') {
+          // Try to make port mappings for all srflx candidates
+          var MAP_LIFETIME = 24 * 60 * 60;  // 24 hours in seconds
+          if (c.type === 'srflx') {
+            if (this.portControl_ === undefined) {
+              log.debug('%1: port control unavailable', this.name_);
+            } else {
+              log.info('%1: port control available', this.name_);
+              this.portControl_.addMapping(c.relatedPort, c.port, MAP_LIFETIME).
+                then((mapping:freedom.PortControl.Mapping) => {
+                  if (mapping.externalPort === -1) {
+                    log.debug('%1: addMapping() failed: %2',
+                        this.name_, mapping);
+                  } else {
+                    log.info('%1: addMapping() success: ',
+                        this.name_, mapping);
+                  }
+              });
+            }
+          }
+        }
+      }
+
+      var churnSignal :ChurnSignallingMessage = {
+        webrtcMessage: message
+      };
+      this.signalForPeerQueue.handle(churnSignal);
+    });
+  }
+
+  private mungeLocalSdp_ = (d:freedom.RTCPeerConnection.RTCSessionDescription)
+      : Promise<freedom.RTCPeerConnection.RTCSessionDescription> => {
+    return this.onceHaveTransformerConfig_.then((tConfig:TransformerConfig) => {
+      if (tConfig.name === 'caesar') {
+        let caesarConfig :caesar.Config = JSON.parse(tConfig.config);
+        d.sdp += 'a=x-uproxy-transform:caesar ' + caesarConfig.key + '\r\n';
```

Comment:
Right now, churn pipe needs two things:
 1. transformer name
 1. transformer config

We could generalise this pretty easily, e.g. by adding two lines to the SDP:
```
a=x-uproxy-transformer-name:caesar
a=x-uproxy-transformer-config:{"key":"25"}
```

When we come to add support for RC4, we'll only have to change your xwalk patch. What do you think? (shouldn't be hard to squeeze JSON configs onto one line safely)